### PR TITLE
Implement `allocate_and_map_memory()`

### PR DESCRIPTION
**Pull request by haraldh**
Allocate memory and map it to a given virtual address.

### CODE REVIEW

**Comment by haraldh** [comment]
```console
/home/harald/.cargo/bin/cargo run --color=always --package enarx-keep-sev-shim --bin enarx-keep-sev-shim
    Finished dev [unoptimized + debuginfo] target(s) in 0.02s
     Running `/home/harald/git/enarx/enarx/enarx-keep-sev-shim/../enarx-keep-sev-test/target/debug/enarx-keep-sev --code ../payload/target/x86_64-unknown-linux-musl/debug/payload --shim target/x86_64-unknown-linux-musl/debug/enarx-keep-sev-shim`
Hypervisor: Starting "target/x86_64-unknown-linux-musl/debug/enarx-keep-sev-shim"
[src/main.rs:82] boot_info = BootInfo {
    setup: Line {
        start: 0,
        end: 1048576,
    },
    code: Line {
        start: 2097152,
        end: 6553600,
    },
    shim: Line {
        start: 1048576,
        end: 1395880,
    },
    mem_size: 2147483648,
}
Hello World!
[src/main.rs:91] &FRAME_ALLOCATOR = BootInfoFrameAllocator {
    max   : 0x0000000080000000,
    next4k: 0x0000000000640000,
    next2M: 0x0000000000800000,
}
init_gdt
[src/main.rs:104] &mem_ptr.as_ptr() = 0x00007f4847fff000
[src/main.rs:105] &mem_ptr[0] = 1
[src/main.rs:106] &mem_ptr[mem_ptr.len() - 1] = 1
[src/main.rs:108] &FRAME_ALLOCATOR = BootInfoFrameAllocator {
    max   : 0x0000000080000000,
    next4k: 0x0000000000a44000,
    next2M: 0x0000000000c00000,
}
[src/kvmvm.rs:346] exit_group(0)
/home/harald/git/enarx/enarx/enarx-keep-sev-test/target/debug/enarx-keep-sev encountered an error:
#0: Handle syscall: ()

Process finished with exit code 1
```

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            let flags = PageTableFlags::PRESENT | PageTableFlags::WRITABLE;
+            unsafe {
+                mapper
+                    .map_to(page, frame, flags, frame_allocator)
+                    .map_err(|_| ())?
+                    .flush();
+            }
+        }
+        Ok(())
+    }
```
Comment (connorkuehl) [inline enarx-keep-sev-shim/src/frame_allocator.rs:146]: I think the same kind of accessibility/namespacing you're going for here can be accomplished by putting `do_allocate_and_map_memory` and `allocate_and_map_memory` into a submodule and only making `allocate_and_map_memory` `pub`. I think doing it that way is a bit easier on the eyes.

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+    let post: &[Page<Size4KiB>] = unsafe {
+        core::slice::from_raw_parts(
+            core::mem::transmute::<_, *const Page<Size4KiB>>(post.as_ptr()),
+            post.len(),
+        )
+    };
+
+    do_allocate_and_map_memory(&pre, mapper, frame_allocator)?;
+    do_allocate_and_map_memory(&middle, mapper, frame_allocator)?;
+    do_allocate_and_map_memory(&post, mapper, frame_allocator)?;
```
Comment (connorkuehl) [inline enarx-keep-sev-shim/src/frame_allocator.rs:187]: I think it's worth a comment here explaining that this is trying to allocate and map using the fewest number of frames possible.

File: enarx-keep-sev-shim/src/main.rs [hunk 3]
```diff
@@ -53,17 +64,49 @@ macro_rules! entry_point {
 
 entry_point!(shim_main);
 
-/// FIXME: will be replaced by a dynamically allocated stack with safe guards
-pub static mut LEVEL_0_STACK: [Page; 5] = [Page::zeroed(); 5];
+/// The global BootInfoFrameAllocator
+pub static mut FRAME_ALLOCATOR: BootInfoFrameAllocator = BootInfoFrameAllocator::empty();
```
Comment (connorkuehl) [inline enarx-keep-sev-shim/src/main.rs:68]: We'll need a lock

File: enarx-keep-sev-shim/src/frame_allocator.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+            let heap_end = heap_start + size - 1u64;
+            let heap_start_page = Page::<S>::containing_address(heap_start);
+            let heap_end_page = Page::<S>::containing_address(heap_end);
+            Page::range_inclusive(heap_start_page, heap_end_page)
+        };
+
+        for page in page_range {
+            let frame = frame_allocator.allocate_frame().ok_or(())?;
+
+            let flags = PageTableFlags::PRESENT | PageTableFlags::WRITABLE;
```
Comment (connorkuehl) [inline enarx-keep-sev-shim/src/frame_allocator.rs:137]: Do we need to worry about any other flags? I.e., kernel vs userspace memory?

File: enarx-keep-sev-shim/src/frame_allocator.rs (hunk 2 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs]: done

File: enarx-keep-sev-shim/src/main.rs (hunk 3 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/main.rs]: was already in the queue :) pushed

File: enarx-keep-sev-shim/src/frame_allocator.rs (hunk 4 above)
Comment (haraldh) [reply enarx-keep-sev-shim/src/frame_allocator.rs]: was already in the queue :) pushed

**Comment by connorkuehl** [comment]
Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) over the singleton macros?

**Comment by haraldh** [comment]
> Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) instead of the singleton macros?

no

**Comment by connorkuehl** [comment]
> > Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) instead of the singleton macros?
> 
> no

Why not?

**Comment by haraldh** [comment]
> Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) instead of the singleton macros?

Some could, but some could lead to a deadlock.

**Comment by haraldh** [comment]
> > > Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) instead of the singleton macros?
> > 
> > no
> 
> Why not?

Also I want to avoid nested Rwlocks for one singleton

**Comment by haraldh** [comment]
> > > Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) instead of the singleton macros?
> > 
> > no
> 
> Why not?

maybe I should give it a try again

**Comment by haraldh** [comment]
> > > > Could we take advantage of [`lazy_static`](https://crates.io/crates/lazy_static) instead of the singleton macros?
> > > 
> > > no
> > 
> > Why not?
> 
> maybe I should give it a try again

Nah... lazy_static is such an overhead for every deref...

**Comment by haraldh** [comment]
Will break it down in smaller chunks

**Pull request closed without merging**